File: src/homework05/models/Phonebook.java
package homework05.models;

import homework05.exceptions.DuplicateContactException;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map.Entry;
import java.util.Objects;

public class Phonebook {
    private final ArrayList<Contact> contacts;
    private final HashMap<String, Contact> contactsMap;

    public Phonebook() {
        contacts = new ArrayList<>();
        contactsMap = new HashMap<>();
    }

    public List<Contact> getContacts() {
        return contacts;
    }

    public void add(Contact contact) throws DuplicateContactException {
        String fullName = contact.fullName();
        if (contactsMap.containsKey(fullName)) throw new DuplicateContactException();

        contacts.add(contact);
        contactsMap.put(fullName, contact);
        contacts.sort(Contact::compareTo);
    }

    public Contact get(int index) {
        Objects.checkIndex(index, this.size());
        return contacts.get(index);
    }

    public void remove(Contact contact) {
        contactsMap.remove(contact.fullName());
        contacts.remove(contact);
    }

    public void remove(int index) {
        Contact contact = contacts.get(index);
        this.remove(contact);
    }

    public List<Contact> find(String name) {
        List<Contact> foundContacts = new ArrayList<>();
        for (Entry<String, Contact> entry : contactsMap.entrySet()) {
            if (entry.getKey().contains(name)) {
                foundContacts.add(entry.getValue());
            }
        }
        return foundContacts;
    }

    public int size() {
        return contacts.size();
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder("Phone book:\n");
        for (Contact contact : contacts) {
            sb.append(contact.toString());
        }
        return sb.toString();
    }
}

File: src/homework07/phonebook/models/phonebook.py
from phonebook.exceptions.phonebook_exceptions import DuplicateContactException


class Phonebook:
    def __init__(self) -> None:
        self._contacts = []
        self._contacts_dict = {}

    @property
    def contacts(self):
        return self._contacts

    @property
    def size(self):
        return len(self._contacts)

    def add(self, contact):
        if contact.full_name in self._contacts_dict:
            raise DuplicateContactException

        self._contacts.append(contact)
        self._contacts_dict[contact.full_name] = contact
        self._contacts.sort()

    def get(self, index):
        return self._contacts[index]

    def remove(self, contact):
        del self._contacts_dict[contact.full_name]
        self._contacts.remove(contact)

    def find(self, name):
        return [v for k, v in self._contacts_dict.items() if k.find(name) >= 0]

    @property
    def phonebook_info(self):
        if self.size == 0:
            return 'Empty phone book'
        s = 'Phone book:\n'
        return s + ''.join(f'{contact.contact_info}\n' for contact in self._contacts)

    def __str__(self) -> str:
        return self.phonebook_info

    def __getitem__(self, item):
        return self._contacts[item]
